lib/compress.js

```diff
@@ -285,6 +285,10 @@ merge(Compressor.prototype, {
 
     AST_Node.DEFMETHOD("reset_opt_flags", function(compressor, rescan) {
         var reduce_vars = rescan && compressor.option("reduce_vars");
+        // Stack of look-up tables to keep track of whether a `SymbolDef` has been
+        // properly assigned before use:
+        // - `push()` & `pop()` when visiting conditional branches
+        // - backup & restore via `save_ids` when visiting out-of-order sections
```

Comment:
@kzc PTAL

Reply:
Fantastic!

Reply:
I always found it strange how the bulk of the work for `reduce_vars` took place in `reset_opt_flags` - before any compress transforms are run - but it makes much more sense to me now. It's essentially static tree analysis that has to be redone with each pass.